src/apprt/gtk-ng/class/command_palette.zig

```diff
@@ -0,0 +1,426 @@
+/// we have to construct a wrapper to be easily consumed by the list model.
+const Command = extern struct {
+    pub const Self = @This();
+    pub const Parent = gobject.Object;
+    parent: Parent,
+
+    pub const getGObjectType = gobject.ext.defineClass(Command, .{
+        .name = "GhosttyCommand",
+        .instanceInit = &init,
+        .classInit = Class.init,
+        .parent_class = &Class.parent,
+        .private = .{ .Type = Private, .offset = &Private.offset },
+    });
+
+    // Expose all fields on the input.Command.C struct as properties that can be
+    // accessed by the GObject type system (and by extension, blueprints)
+    const properties = props: {
+        const info = @typeInfo(input.Command.C).@"struct";
+        var props: [info.fields.len]type = undefined;
+
+        for (info.fields, 0..) |field, i| {
+            const accessor = struct {
+                fn getter(self: *Self) ?[:0]const u8 {
+                    const priv = self.private();
+                    return std.mem.span(@field(priv.cmd_c, field.name));
+                }
+            };
+
+            // "Canonicalize" field names into the format GObject expects
+            const prop_name = prop_name: {
+                var buf: [field.name.len:0]u8 = undefined;
+                _ = std.mem.replace(u8, field.name, "_", "-", &buf);
+                break :prop_name buf;
+            };
+
+            props[i] = gobject.ext.defineProperty(
+                &prop_name,
+                Command,
+                ?[:0]const u8,
+                .{
+                    .default = null,
+                    .accessor = gobject.ext.typedAccessor(
+                        Command,
+                        ?[:0]const u8,
+                        .{
+                            .getter = &accessor.getter,
+                        },
+                    ),
+                },
+            );
+        }
+
+        break :props props;
+    };
+
+    pub const Private = struct {
+        /// The command.
+        cmd_c: input.Command.C,
+
+        pub var offset: c_int = 0;
+    };
+
+    pub fn new(alloc: Allocator, cmd: input.Command, keybinds: input.Binding.Set) (Allocator.Error || error{NoSpaceLeft})!*Command {
+        const self = gobject.ext.newInstance(Command, .{});
+        const priv = self.private();
+
+        var buf: [64]u8 = undefined;
+
+        const action = action: {
+            const trigger = keybinds.getTrigger(cmd.action) orelse break :action null;
+            const accel = try key.accelFromTrigger(&buf, trigger) orelse break :action null;
+            break :action try alloc.dupeZ(u8, accel);
+        };
+
+        priv.cmd_c = .{
+            .title = cmd.title.ptr,
+            .description = cmd.description.ptr,
+            .action = if (action) |v| v.ptr else "",
+            .action_key = try std.fmt.allocPrintZ(alloc, "{}", .{cmd.action}),
+        };
```

Comment:
Same thoughts here around `new` vs `init`. And I wonder if we should make the C type a boxed type so we can just pass it in as a normal GTK property.

Reply:
Well it can't be a boxed type because the gio.ListModel type is annoying and requires its items to implement gobject.Object 😅